internal/function_time_parser.go

```diff
@@ -584,61 +612,54 @@ func monthDayYearFormatter(t *time.Time) ([]rune, error) {
 }
 
 func dayParser(text []rune, t *time.Time) (int, error) {
-	const dayLen = 2
-	if len(text) < dayLen {
-		return 0, fmt.Errorf("unexpected day number")
-	}
-	d, err := strconv.ParseInt(string(text[:dayLen]), 10, 64)
+	progress, days, err := parseDigitRespectingOptionalPlaces(text, 1, 31)
 	if err != nil {
-		return 0, fmt.Errorf("unexpected day number")
-	}
-	if d < 0 {
-		return 0, fmt.Errorf("invalid day number %d", d)
+		return 0, fmt.Errorf("could not parse day number: %s", err)
 	}
 	*t = time.Date(
 		int(t.Year()),
 		t.Month(),
-		int(d),
+		int(days),
 		int(t.Hour()),
 		int(t.Minute()),
 		int(t.Second()),
 		int(t.Nanosecond()),
 		t.Location(),
 	)
-	return dayLen, nil
+	return progress, nil
 }
 
 func dayFormatter(t *time.Time) ([]rune, error) {
 	return []rune(fmt.Sprintf("%02d", t.Day())), nil
 }
 
 func yearMonthDayParser(text []rune, t *time.Time) (int, error) {
-	fmtLen := len("2021-01-20")
-	if len(text) < fmtLen {
-		return 0, fmt.Errorf("unexpected year-month-day format")
-	}
-	splitted := strings.Split(string(text[:fmtLen]), "-")
-	if len(splitted) != 3 {
-		return 0, fmt.Errorf("unexpected year-month-day format")
+	const separator = '-'
+	progress, y, err := parseDigitRespectingOptionalPlaces(text, 1, 9999)
+	if err != nil {
+		return 0, fmt.Errorf("could not parse year number: %s", err)
 	}
-	year := splitted[0]
-	month := splitted[1]
-	day := splitted[2]
-	if len(year) != 4 || len(month) != 2 || len(day) != 2 {
-		return 0, fmt.Errorf("unexpected year-month-day format")
+	if len(text) < progress || text[progress] != separator {
+		return 0, fmt.Errorf("could not parse year-month-day: [%c] not found after [%s]", separator, string(text[:progress]))
 	}
-	y, err := strconv.ParseInt(year, 10, 64)
+	progress += 1
+
+	mProgress, m, err := parseDigitRespectingOptionalPlaces(text[progress:], 1, 12)
 	if err != nil {
-		return 0, fmt.Errorf("unexpected year-month-day format: %w", err)
+		return 0, fmt.Errorf("could not parse month number: %s", err)
 	}
-	m, err := strconv.ParseInt(month, 10, 64)
-	if err != nil {
-		return 0, fmt.Errorf("unexpected year-month-day format: %w", err)
+	progress += mProgress
+	if len(text) < progress || text[progress] != separator {
```

Comment:
`len(text) == progress` is also out of boundary. So we need to check `len(text) == progress`.
Furthermore, `text[:progress]` is not available in this case, so it must be `text` only.